### ISSUE TITLE
the first question

### ISSUE THREAD

**Issue by kwea123** [oxford-cs-deepnlp-2017/practical-2#1]
Hello,

My name is Quei-An, an Msc student. For the first question of practical 2, I don't understand much what "starting from random embeddings" means. You mention GloVe afterwards, so I suppose that stands for word embeddings, but how can we train all word embeddings with such limited amount of data (only ~2000 talks)?

I tried somehow to implement the model with respect to the classification problem, but firstly I can only feed in one input at a time because each document has different size (number of words), secondly there's so little data that the accuracy turns out to be so poor. Then you mention "Training in batches", so I start getting confused. 

Could you please clarify?

If I use fixed word representations (e.g. word2vec) then everything seems OK, I get about 56% accuracy by trying different hyperparameters.

Thank you and have a nice weekend.

**Comment by hmate9** [comment]
In the previous exercise we trained Word2Vec from scratch:

`from gensim.models import Word2Vec
model_ted = Word2Vec(sentences_ted, min_count=10)`

I haven't done the following, but I assume starting from GloVe means we start the training process where the embedding model is initialized to the weights of an already trained model that was downloaded.

"Training in batches" is a way to optimize the speed of the training algorithm. Say your batch size is 50. That means you run 50 samples through your network before updating the weights. 

Your accuracy of 56% seems a little low. As a benchmark, I get 85% with a super simple NN that has layer sizes of 32, 16 and 8 for the final output layer. 

54% of the talks do not contain "technology", "entertainment" or "design". So I think maybe what is happening is that your NN is basically predicting "none" category every time and not actually improving.

**Comment by kwea123** [comment]
Yes but the question asks us to consider 3 scenarios : starting from random embeddings, starting from GloVe embeddings or fixed to be the GloVe values. 

In the first two scenarios, I think we should also consider word embedding vectors as parameters, so that makes about 50k words *100 embedding size = 5m parameters in addition. I don't know how the training  procedure can update all these values with only ~2000 data properly, that's my question. And my problem with batches is that since each document contains different number of words, I can't train them in batches because tensorflow doesn't accept tensors with different sizes (I don't know if other platforms do) so I can only feed in one input at a time.

Anyway, if I use fixed vector representation (we don't update them while training), then I train in batches of course, but still that doesn't give me so good results as you said. I use word2vec cbow model(default) trained on the same talks instead of GloVe, and I also tried your NN structure with activations relu,relu and softmax (without dropouts) but it still gives me the same 56%. Might the word representation be the problem?

PS. I get 83% for multilabel classification, not sure what makes the difference

**Comment by rhaps0dy** [comment]
> 50k words *100 embedding size = 5m parameters in addition. I don't know how the training procedure can update all these values with only ~2000 data properly, that's my question

It's probably a way to make us experience our classifier overfitting. "Why does my validation accuracy go down?", I wondered for a while.

> PS. I get 83% for multilabel classification, not sure what makes the difference

Possibly the way accuracy is computed. In single-label classification, not getting the exact label (say, outputting oox instead of oxx) counts as 0. Instead, in multi-label classification, it counts as 1/3.

**Comment by iassael** [comment]
Well said. For this practical, the variable number of words is tackled by using an average pooling (mean) layer, over the sequence dimension. For PyTorch you could have a look at nn.Embedding and this example of MapTable https://github.com/amdegroot/pytorch-containers#maptable coupled with an AvgPool layer (similarly for TF). Is it more clear now?

**Comment by bshillingford** [comment]
To add to that, there's more than one way . Let _E_ be an embedding matrix, so that _E x_ selects one column of _E_. If we want the sum of embeddings for _n_ words: _E x1 + E x2 + ... + E xn_, note this is equal to _E(x1 + x2 + ... + xn)_. Since a sum of one-hot vectors is a vector of counts of words, we can simply produce a counts vector and multiply this by _E_ to produce an identical result.

If you're using a framework that has an efficient operation to multiply sparse vectors by dense matrices, then this is clearly more efficient than producing all the embeddings them summing them. In other situations it may be interesting to adapt this to alternative fast approaches.

**Comment by kwea123** [comment]
Yes I came up with this idea last night and am implementing it. Thank you for all your comments, it's more clear.
